controllers/backup/backup_controller.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2022 The SODA Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backup
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/soda-cdm/kahu/apis/kahu/v1beta1"
+	kahuclientset "github.com/soda-cdm/kahu/client/clientset/versioned"
+	kahuv1client "github.com/soda-cdm/kahu/client/clientset/versioned/typed/kahu/v1beta1"
+	kinf "github.com/soda-cdm/kahu/client/informers/externalversions/kahu/v1beta1"
+	kahulister "github.com/soda-cdm/kahu/client/listers/kahu/v1beta1"
+	"github.com/soda-cdm/kahu/controllers"
+	"github.com/soda-cdm/kahu/controllers/backup/cmd/options"
+	metaservice "github.com/soda-cdm/kahu/providerframework/meta_service/lib/go"
+	utils "github.com/soda-cdm/kahu/utils"
+)
+
+var (
+	grpcServer     *grpc.Server
+	grpcConnection *grpc.ClientConn
+	metaClient     metaservice.MetaServiceClient
+)
+
+type Controller struct {
+	*controllers.BaseController
+	client       kubernetes.Interface
+	klient       kahuclientset.Interface
+	backupSynced []cache.InformerSynced
+	kLister      kahulister.BackupLister
+	config       *restclient.Config
+	bkpClient    kahuv1client.BackupsGetter
+	bkpLocClient kahuv1client.BackupLocationsGetter
+	flags        *options.BackupControllerFlags
+}
+
+func NewController(klient kahuclientset.Interface,
+	backupInformer kinf.BackupInformer,
+	config *restclient.Config,
+	bkpClient kahuv1client.BackupsGetter,
+	bkpLocClient kahuv1client.BackupLocationsGetter,
+	flags *options.BackupControllerFlags) *Controller {
+
+	c := &Controller{
+		BaseController: controllers.NewBaseController(controllers.Backup),
+		klient:         klient,
+		kLister:        backupInformer.Lister(),
+		config:         config,
+		bkpClient:      bkpClient,
+		bkpLocClient:   bkpLocClient,
+		flags:          flags,
+	}
+
+	backupInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.handleAdd,
+			DeleteFunc: c.handleDel,
+		},
+	)
+
+	return c
+}
+
+func (c *Controller) Run(ch chan struct{}) error {
```

Comment:
IMO, ```Run```, ```worker``` and ```processNextItem``` can be part of common controller

Reply:
Created issues to track this #43